app/controllers/devise_sessions_controller.rb

```diff
@@ -3,54 +3,78 @@ class DeviseSessionsController < Devise::SessionsController
+
+    @email = params.dig(:user, :email)
+    catch(:warden) do
+      self.resource = warden.authenticate(auth_options)
+    end
 
-    self.resource = warden.authenticate(auth_options)
     if resource
-      login_state.update!(password_ok: true)
+      user = User.find_by(email: @email).id
+      @login_state = LoginState.create!(
+        created_at: Time.zone.now,
+        user_id: user,
+        redirect_path: after_login_path(payload, user),
+        jwt_id: session[:jwt_id],
+      )
+
+      @login_state_id = login_state.id
+      session[:login_state_id] = login_state.id
+
       if request.env["warden.mfa.required"]
         MultiFactorAuth.generate_and_send_code(resource)
         redirect_to user_session_phone_code_path
       else
         do_sign_in
       end
     else
-      @password_error_message =
-        case User.find_by(email: params.dig(:user, :email))&.unauthenticated_message
-        when :last_attempt
-          I18n.t("devise.failure.last_attempt")
-        when :locked
-          I18n.t("devise.failure.locked")
-        else
-          I18n.t("devise.sessions.new.fields.password.errors.incorrect")
-        end
-      render :new
-    end
-  end
+      @resource_error_messages = {}
 
-  def phone_code; end
+      @resource_error_messages[:email] = if @email.blank?
+                                           [I18n.t("activerecord.errors.models.user.attributes.email.blank")]
+                                         elsif !Devise.email_regexp.match? @email
+                                           [I18n.t("activerecord.errors.models.user.attributes.email.invalid")]
+                                         end
 
-  def phone_code_send
-    MultiFactorAuth.generate_and_send_code(login_state.user)
+      if params.dig(:user, :password).blank?
+        @resource_error_messages[:password] = [I18n.t("activerecord.errors.models.user.attributes.password.blank")]
+      end
+
+      user = User.find_by(email: @email)
+      user_exists = user.present?
+
+      if user_exists && !user.active_for_authentication?
+        @resource_error_messages[:email] = [I18n.t("devise.failure.unconfirmed")]
+      end
+
+      if @email.present? && params.dig(:user, :password).present? && user_exists
+        authentication_failure = user&.unauthenticated_message # pragma: allowlist secrets
+        @resource_error_messages[:password] = [I18n.t("devise.failure.#{authentication_failure}")] if authentication_failure
+      elsif @email.present? && params.dig(:user, :password).blank? && user_exists
+        @resource_error_messages[:password] = [I18n.t("activerecord.errors.models.user.attributes.password.blank")]
+      elsif @email.present?
+        redirect_to new_user_registration_start_path(user: { email: @email }) and return if payload
+
+        render "devise/registrations/transition_checker" and return if Rails.configuration.force_jwt_at_registration
+
+        @resource_error_messages[:email] = [I18n.t("devise.failure.no_account")]
```

Comment:
This will overwrite the previous `@resouce_error_messages[:email] = [I18n.t("activerecord.errors.models.user.attributes.email.invalid")]`, is that right?

Reply:
Good spot. I've moved the blank and invalid errors down into an `elsif`.